src/tree/updater_quantile_hist.cc

```diff
@@ -156,12 +156,18 @@ void QuantileHistMaker::Builder::Update(const GHistIndexMatrix& gmat,
       const int cright = (*p_tree)[nid].RightChild();
       hist_.AddHistRow(cleft);
       hist_.AddHistRow(cright);
-      if (row_set_collection_[cleft].Size() < row_set_collection_[cright].Size()) {
+      if (rabit::IsDistributed()) {
+        // in distributed mode, we need to keep consistent across workers
```

Comment:
I see what you mean

Given the several facts

1. the benefit comes when there is data skew in each node (actually not in Spark case, Spark's repartition is essentially round-robin style) or the max value in left/right across workers has skew

2. the network overhead is high with the current rabit implementation

I think we can make it as an option with three different values (always from left, always from right, communicate to get max_num for each node)

what do you think

Reply:
There is no need to prematurely optimise, I mention it because it had an impact for me in a multi-GPU setting. This may not be true for you and you could see how it performs and do a future PR.